File: sample/cpp/13_maze_generation_steps.cpp
#include "runtime/cpp/native/core/py_runtime.h"
#include "runtime/cpp/native/core/process_runtime.h"
#include "runtime/cpp/native/core/scope_exit.h"

#include "generated/built_in/io_ops.h"
#include "generated/built_in/sequence.h"
#include "generated/std/time.h"
#include "generated/utils/gif.h"

// 13: Sample that outputs DFS maze-generation progress as a GIF.

bytes capture(const rc<list<list<int64>>>& grid, int64 w, int64 h, int64 scale) {
    int64 width = w * scale;
    int64 height = h * scale;
    bytearray frame = bytearray(width * height);
    for (int64 y = 0; y < h; ++y) {
        for (int64 x = 0; x < w; ++x) {
            int64 v = (py_list_at_ref(rc_list_ref(py_list_at_ref(rc_list_ref(grid), py_to<int64>(y))), py_to<int64>(x)) == 0 ? 255 : 40);
            for (int64 yy = 0; yy < scale; ++yy) {
                int64 base = (y * scale + yy) * width + x * scale;
                for (int64 xx = 0; xx < scale; ++xx)
                    frame[base + xx] = v;
            }
        }
    }
    return frame;
}

void run_13_maze_generation_steps() {
    // Increase maze size and render resolution to ensure sufficient runtime.
    int64 cell_w = 89;
    int64 cell_h = 67;
    int64 scale = 5;
    int64 capture_every = 20;
    str out_path = "sample/out/13_maze_generation_steps.gif";
    
    float64 start = pytra::std::time::perf_counter();
    rc<list<list<int64>>> grid = rc_list_from_value(list<list<int64>>(cell_h, list<int64>(cell_w, 1)));
    rc<list<::std::tuple<int64, int64>>> stack = rc_list_from_value(list<::std::tuple<int64, int64>>{::std::make_tuple(1, 1)});
    py_list_at_ref(py_list_at_ref(rc_list_ref(grid), py_to<int64>(1)), py_to<int64>(1)) = 0;
    
    rc<list<::std::tuple<int64, int64>>> dirs = rc_list_from_value(list<::std::tuple<int64, int64>>{::std::make_tuple(2, 0), ::std::make_tuple(-(2), 0), ::std::make_tuple(0, 2), ::std::make_tuple(0, -(2))});
    rc<list<bytes>> frames = rc_list_from_value(list<bytes>{});
    int64 step = 0;
    
    while (!((rc_list_ref(stack)).empty())) {
        auto [x, y] = py_list_at_ref(rc_list_ref(stack), py_to<int64>(-(1)));
        rc<list<::std::tuple<int64, int64, int64, int64>>> candidates = rc_list_from_value(list<::std::tuple<int64, int64, int64, int64>>{});
        for (int64 k = 0; k < 4; ++k) {
            auto [dx, dy] = py_list_at_ref(rc_list_ref(dirs), py_to<int64>(k));
            int64 nx = x + dx;
            int64 ny = y + dy;
            if ((nx >= 1) && (nx < cell_w - 1) && (ny >= 1) && (ny < cell_h - 1) && (py_list_at_ref(rc_list_ref(py_list_at_ref(rc_list_ref(grid), py_to<int64>(ny))), py_to<int64>(nx)) == 1)) {
                if (dx == 2) {
                    py_list_append_mut(rc_list_ref(candidates), ::std::make_tuple(nx, ny, x + 1, y));
                } else if (dx == -(2)) {
                    py_list_append_mut(rc_list_ref(candidates), ::std::make_tuple(nx, ny, x - 1, y));
                } else if (dy == 2) {
                    py_list_append_mut(rc_list_ref(candidates), ::std::make_tuple(nx, ny, x, y + 1));
                } else {
                    py_list_append_mut(rc_list_ref(candidates), ::std::make_tuple(nx, ny, x, y - 1));
                }
            }
        }
        if ((rc_list_ref(candidates)).empty()) {
            py_list_pop_mut(rc_list_ref(stack));
        } else {
            auto __idx_3 = (x * 17 + y * 29 + (rc_list_ref(stack)).size() * 13) % (rc_list_ref(candidates)).size();
            ::std::tuple<int64, int64, int64, int64> sel = py_list_at_ref(rc_list_ref(candidates), py_to<int64>(__idx_3));
            auto [nx, ny, wx, wy] = sel;
            py_list_at_ref(py_list_at_ref(rc_list_ref(grid), py_to<int64>(wy)), py_to<int64>(wx)) = 0;
            py_list_at_ref(py_list_at_ref(rc_list_ref(grid), py_to<int64>(ny)), py_to<int64>(nx)) = 0;
            py_list_append_mut(rc_list_ref(stack), ::std::make_tuple(nx, ny));
        }
        if (step % capture_every == 0)
            py_list_append_mut(rc_list_ref(frames), capture(grid, cell_w, cell_h, scale));
        step++;
    }
    py_list_append_mut(rc_list_ref(frames), capture(grid, cell_w, cell_h, scale));
    pytra::utils::gif::save_gif(out_path, cell_w * scale, cell_h * scale, rc_list_ref(frames), pytra::utils::gif::grayscale_palette(), 4, 0);
    float64 elapsed = pytra::std::time::perf_counter() - start;
    py_print("output:", out_path);
    py_print("frames:", (rc_list_ref(frames)).size());
    py_print("elapsed_sec:", elapsed);
}

int main(int argc, char** argv) {
    pytra_configure_from_argv(argc, argv);
    run_13_maze_generation_steps();
    return 0;
}

File: sample/java/13_maze_generation_steps.java
final class _13_maze_generation_steps {
    private _13_maze_generation_steps() {
    }


    // 13: Sample that outputs DFS maze-generation progress as a GIF.

    public static java.util.ArrayList<Long> capture(java.util.ArrayList<java.util.ArrayList<Long>> grid, long w, long h, long scale) {
        long width = w * scale;
        long height = h * scale;
        java.util.ArrayList<Long> frame = PyRuntime.__pytra_bytearray(width * height);
        for (long y = 0L; y < h; y += 1L) {
            for (long x = 0L; x < w; x += 1L) {
                long v = ((((((Long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).get((int)((((x) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).size())) + (x)) : (x))))))) == (0L))) ? (255L) : (40L));
                for (long yy = 0L; yy < scale; yy += 1L) {
                    long base = (y * scale + yy) * width + x * scale;
                    for (long xx = 0L; xx < scale; xx += 1L) {
                        frame.set((int)((((base + xx) < 0L) ? (((long)(frame.size())) + (base + xx)) : (base + xx))), v);
                    }
                }
            }
        }
        return PyRuntime.__pytra_bytearray(frame);
    }

    public static void run_13_maze_generation_steps() {
        long cell_w = 89L;
        long cell_h = 67L;
        long scale = 5L;
        long capture_every = 20L;
        String out_path = "sample/out/13_maze_generation_steps.gif";
        double start = time.perf_counter();
        java.util.ArrayList<java.util.ArrayList<Long>> grid = new java.util.ArrayList<java.util.ArrayList<Long>>();
        for (long __ = 0L; __ < cell_h; __ += 1L) {
            grid.add(PyRuntime.__pytra_list_repeat(1L, cell_w));
        }
        java.util.ArrayList<java.util.ArrayList<Object>> stack = new java.util.ArrayList<java.util.ArrayList<Object>>(java.util.Arrays.asList(new java.util.ArrayList<Object>(java.util.Arrays.asList(1L, 1L))));
        ((java.util.ArrayList<Object>)(Object)(grid.get((int)((((1L) < 0L) ? (((long)(grid.size())) + (1L)) : (1L)))))).set((int)((((1L) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((1L) < 0L) ? (((long)(grid.size())) + (1L)) : (1L)))))).size())) + (1L)) : (1L))), 0L);
        java.util.ArrayList<java.util.ArrayList<Object>> dirs = new java.util.ArrayList<java.util.ArrayList<Object>>(java.util.Arrays.asList(new java.util.ArrayList<Object>(java.util.Arrays.asList(2L, 0L)), new java.util.ArrayList<Object>(java.util.Arrays.asList((-(2L)), 0L)), new java.util.ArrayList<Object>(java.util.Arrays.asList(0L, 2L)), new java.util.ArrayList<Object>(java.util.Arrays.asList(0L, (-(2L))))));
        java.util.ArrayList<java.util.ArrayList<Long>> frames = new java.util.ArrayList<java.util.ArrayList<Long>>();
        long step = 0L;
        while (((stack) != null && !(stack).isEmpty())) {
            java.util.ArrayList<Object> __tuple_0 = ((java.util.ArrayList<Object>)(Object)(stack.get((int)(((((-(1L))) < 0L) ? (((long)(stack.size())) + ((-(1L)))) : ((-(1L))))))));
            long x = ((Long)(__tuple_0.get(0)));
            long y = ((Long)(__tuple_0.get(1)));
            java.util.ArrayList<java.util.ArrayList<Object>> candidates = new java.util.ArrayList<java.util.ArrayList<Object>>();
            long nx = 0L;
            long ny = 0L;
            for (long k = 0L; k < 4L; k += 1L) {
                java.util.ArrayList<Object> __tuple_1 = ((java.util.ArrayList<Object>)(Object)(dirs.get((int)((((k) < 0L) ? (((long)(dirs.size())) + (k)) : (k))))));
                long dx = ((Long)(__tuple_1.get(0)));
                long dy = ((Long)(__tuple_1.get(1)));
                nx = x + dx;
                ny = y + dy;
                if ((((nx) >= (1L)) && ((nx) < (cell_w - 1L)) && ((ny) >= (1L)) && ((ny) < (cell_h - 1L)) && ((((Long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((ny) < 0L) ? (((long)(grid.size())) + (ny)) : (ny)))))).get((int)((((nx) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((ny) < 0L) ? (((long)(grid.size())) + (ny)) : (ny)))))).size())) + (nx)) : (nx))))))) == (1L)))) {
                    if (((dx) == (2L))) {
                        candidates.add(new java.util.ArrayList<Object>(java.util.Arrays.asList(nx, ny, x + 1L, y)));
                    } else {
                        if (((dx) == ((-(2L))))) {
                            candidates.add(new java.util.ArrayList<Object>(java.util.Arrays.asList(nx, ny, x - 1L, y)));
                        } else {
                            if (((dy) == (2L))) {
                                candidates.add(new java.util.ArrayList<Object>(java.util.Arrays.asList(nx, ny, x, y + 1L)));
                            } else {
                                candidates.add(new java.util.ArrayList<Object>(java.util.Arrays.asList(nx, ny, x, y - 1L)));
                            }
                        }
                    }
                }
            }
            if (((((long)(candidates.size()))) == (0L))) {
                stack.remove(stack.size() - 1);
            } else {
                Object sel = candidates.get((int)(((((x * 17L + y * 29L + ((long)(stack.size())) * 13L) % ((long)(candidates.size()))) < 0L) ? (((long)(candidates.size())) + ((x * 17L + y * 29L + ((long)(stack.size())) * 13L) % ((long)(candidates.size())))) : ((x * 17L + y * 29L + ((long)(stack.size())) * 13L) % ((long)(candidates.size()))))));
                java.util.ArrayList<Object> __tuple_2 = ((java.util.ArrayList<Object>)(Object)(sel));
                nx = ((Long)(__tuple_2.get(0)));
                ny = ((Long)(__tuple_2.get(1)));
                long wx = ((Long)(__tuple_2.get(2)));
                long wy = ((Long)(__tuple_2.get(3)));
                ((java.util.ArrayList<Object>)(Object)(grid.get((int)((((wy) < 0L) ? (((long)(grid.size())) + (wy)) : (wy)))))).set((int)((((wx) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((wy) < 0L) ? (((long)(grid.size())) + (wy)) : (wy)))))).size())) + (wx)) : (wx))), 0L);
                ((java.util.ArrayList<Object>)(Object)(grid.get((int)((((ny) < 0L) ? (((long)(grid.size())) + (ny)) : (ny)))))).set((int)((((nx) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((ny) < 0L) ? (((long)(grid.size())) + (ny)) : (ny)))))).size())) + (nx)) : (nx))), 0L);
                stack.add(new java.util.ArrayList<Object>(java.util.Arrays.asList(nx, ny)));
            }
            if (((step % capture_every) == (0L))) {
                frames.add(capture(grid, cell_w, cell_h, scale));
            }
            step += 1L;
        }
        frames.add(capture(grid, cell_w, cell_h, scale));
        gif.save_gif(out_path, cell_w * scale, cell_h * scale, frames, gif.grayscale_palette(), 4L, 0L);
        double elapsed = time.perf_counter() - start;
        System.out.println(String.valueOf("output:") + " " + String.valueOf(out_path));
        System.out.println(String.valueOf("frames:") + " " + String.valueOf(((long)(frames.size()))));
        System.out.println(String.valueOf("elapsed_sec:") + " " + String.valueOf(elapsed));
    }
}
